sd/internal/instance/cache.go

```diff
@@ -81,3 +86,20 @@ func (r registry) register(c chan<- sd.Event) {
 func (r registry) deregister(c chan<- sd.Event) {
 	delete(r, c)
 }
+
+// copyEvent does a deep copy on sd.Event
+func copyEvent(e sd.Event) sd.Event {
+	// observers all need their own copy of event
+	// because they can directly modify event.Instances
+	// for example, by calling sort.Strings
+	var instances []string
+	if len(e.Instances) != 0 {
+		instances = make([]string, len(e.Instances))
+		copy(instances, e.Instances)
+	}
+	eventCopy := sd.Event{
```

Comment:
Maybe it is not important to new a struct here, `e` is already a copy, just `e.Instances=instances` is the same.

Reply:
Thanks for the thorough review, I've improved the implementation of copyEvent to also handle another potential edge case where a 0 len slice with some capacity could be shared and later appended to causing data races inside the pre-allocated memory.